### ISSUE TITLE
Explore technical challenges with integrating with Feast

### ISSUE THREAD

**Issue by kimwnasptd** [canonical/bundle-kubeflow#1064]
### Context

We want to evaluate integrating [Feast](https://docs.feast.dev/v/master/getting-started/architecture/overview) with CKF, for future CKF releases.

Some use-cases include but are not limited to:
- Feature Engineering (usually via a Notebook)
- Online Serving of features

The goal of this task is to explore:
1. The Software pieces and integrations to acomodate the use of Feast
2. Potential technical architectures with Charms, that can allow for the agreed use-cases

This task is aimed to be ground up exploration of the technical aspects of Feast. To have a full proposal we'll also need to better define the use-cases. But the use cases will be out of scope for this task, although will need to be done in parallel to ensure the technical approaches handle the necessary user stories.

### What needs to get done

1. Look into and understand the [architecture of Feast](https://docs.feast.dev/v/master/getting-started/architecture/overview) and how to [deploy in production](https://docs.feast.dev/v/master/how-to-guides/running-feast-in-production)
2. Ideally, have a small demo of using Feast from a K8s cluster (it could use both charms and upstream manifests)

### Definition of Done

1. We know how to [run Feast in production](https://docs.feast.dev/v/master/how-to-guides/running-feast-in-production#summary)
2. We have a description of what we would need to deploy and relate with what Charms to support such a production deployment of Feast

**Comment by misohu** [comment]
As a demo for the kick off meeting me and @NohaIhab have created an [advanced example](https://github.com/misohu/feast-docker-compose) for feast in docker compose where  we've extended the official [Feast Feature Store demo](https://docs.feast.dev/getting-started/quickstart) by introducing several key features and infrastructure improvements:

1. **PostgreSQL as Offline Store**: 
   - The offline store now uses PostgreSQL, allowing for more scalable and reliable storage of historical feature data.
   - The data pipeline includes a notebook (`load-data.ipynb`) to load the sample data (`driver_stats.parquet`) into the offline store.

2. **Redis as Online Store**: 
   - Redis has been integrated as the online store for serving real-time features. This setup improves performance for serving features in low-latency scenarios.

3. **PostgreSQL for Feature Registry**: 
   - I've added a separate PostgreSQL instance specifically for storing the Feast feature registry. This stores feature definitions and metadata to keep track of all registered features.

4. **Custom Feast UI Container**:
   - A custom Docker container is built from the `feast_ui` directory to run the Feast UI on port `8889`. This allows users to interact with their feature store through a browser-based UI.
   - The container installs the necessary Feast dependencies (`feast`, `feast[postgres]`) and is integrated with the feature store and registry.

5. **Two Configurations for Feature Store**:
   - **`feature_store.yaml`**: This file is used for local development (e.g., when accessing Feast from a Jupyter notebook).
   - **`feature_store_compose.yaml`**: This configuration is mounted into the Feast UI container and used in the Docker Compose setup.

6. **Docker Compose Setup**:
   - The Docker Compose file orchestrates the services, including PostgreSQL, Redis, and the custom Feast UI container. The services are configured to work together seamlessly, and the setup allows easy local deployment and testing.
   - Running `docker-compose up --build` starts all services and makes the Feast UI accessible at `http://localhost:8889`.

7. **Jupyter Notebooks**:
   - The `demo.ipynb` notebook demonstrates how to interact with the feature store.
   - The project structure includes example notebooks and configuration files for loading data, accessing the feature store, and testing the workflow.

This setup provides a robust environment for exploring and interacting with Feast using Docker containers, with support for both online and offline stores, as well as a custom UI.

**Comment by kimwnasptd** [comment]
@misohu that's awesome context!

So then what would be a potential plan for charming parts of the workflow? The immediate ones that come to mind are:
1. Charm the Feast UI
2. Use the PostgreSQL and Reddis charms for the offline/online stores

But what about a charm that could somehow help with the `feature_store.yaml` file? An example idea would be to have such a charm that expresses the offline and online store as relations to the DB Charms, and potentially have an action that prints then the final `feature_store.yaml` that a user could use?

**Comment by misohu** [comment]
Hey @kimwnasptd on the sync meeting we were talking about similar design as this 

![feast drawio](https://github.com/user-attachments/assets/98ce2c65-beed-4e50-8a6b-e974be302eab)

Now I personally believe that most of the times the user already has a db for offline store with data so the design can also keep in mind the option to make the offline store configurable for outside database.  We can also make the online store configurable outside of the charms.
